receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go

```diff
@@ -0,0 +1,290 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package processscraper
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"runtime"
+	"time"
+
+	"github.com/shirou/gopsutil/cpu"
+	"github.com/shirou/gopsutil/host"
+	"github.com/shirou/gopsutil/process"
+	"go.opencensus.io/trace"
+
+	"go.opentelemetry.io/collector/component/componenterror"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/internal/processor/filterset"
+)
+
+// scraper for Process Metrics
+type scraper struct {
+	config    *Config
+	startTime pdata.TimestampUnixNano
+	includeFS filterset.FilterSet
+	excludeFS filterset.FilterSet
+}
+
+type processMetadata struct {
+	pid      int32
+	name     string
+	handle   processHandle
+	username string
+	cmdline  string
+}
+
+type processTimes struct {
+	*processMetadata
+	times *cpu.TimesStat
+}
+
+type processMemoryInfo struct {
+	*processMetadata
+	memoryInfo *process.MemoryInfoStat
+}
+
+type processIoCounters struct {
+	*processMetadata
+	ioCounters *process.IOCountersStat
+}
+
+// newProcessScraper creates a Process Scraper
+func newProcessScraper(_ context.Context, cfg *Config) (*scraper, error) {
+	scraper := &scraper{config: cfg}
+
+	var err error
+
+	if len(cfg.Include.Names) > 0 {
+		scraper.includeFS, err = filterset.CreateFilterSet(cfg.Include.Names, &cfg.Include.Config)
+		if err != nil {
+			return nil, fmt.Errorf("error creating process include filters: %v", err)
+		}
+	}
+
+	if len(cfg.Exclude.Names) > 0 {
+		scraper.excludeFS, err = filterset.CreateFilterSet(cfg.Exclude.Names, &cfg.Exclude.Config)
+		if err != nil {
+			return nil, fmt.Errorf("error creating process exclude filters: %v", err)
+		}
+	}
+
+	return scraper, nil
+}
+
+// Initialize
+func (s *scraper) Initialize(_ context.Context) error {
+	bootTime, err := host.BootTime()
+	if err != nil {
+		return err
+	}
+
+	s.startTime = pdata.TimestampUnixNano(bootTime)
+	return nil
+}
+
+// Close
+func (s *scraper) Close(_ context.Context) error {
+	return nil
+}
+
+// ScrapeMetrics
+func (s *scraper) ScrapeMetrics(ctx context.Context) (pdata.MetricSlice, error) {
+	_, span := trace.StartSpan(ctx, "processscraper.ScrapeMetrics")
+	defer span.End()
+
+	var errors []error
+	metrics := pdata.NewMetricSlice()
+
+	processes, err := s.getProcesses()
+	if len(processes) == 0 {
+		return metrics, err
+	}
+
+	if err != nil {
+		errors = append(errors, err)
+	}
+
+	cpuTimes := make([]*processTimes, 0, len(processes))
+	memoryInfo := make([]*processMemoryInfo, 0, len(processes))
+	ioCounters := make([]*processIoCounters, 0, len(processes))
+
+	for _, metadata := range processes {
+		times, err := metadata.handle.Times()
+		if err != nil {
+			errors = append(errors, fmt.Errorf("error reading process cpu times for process %q (pid %v): %v", metadata.name, metadata.pid, err))
```

Comment:
Shouldn't we append the empty / default item to `cpuTimes` here to preserve order of the processes?
AFAICT the index of items in each slice must exactly match the index of process in `processes`.

Reply:
The ordering of data in these slices doesn't matter, but the code was kinda hard to follow since we were obtaining the data for three different metrics and then initializaing them later. I've refactored this so each metric is scraped & initialized one-by-one, and its hopefully much easier to follow now.